ragna/deploy/_ui/components/metadata_filters_builder.py

```diff
@@ -0,0 +1,248 @@
+            return
+
+        self.value_select.disabled = False
+        self.operator_select.disabled = False
+        self.param.value.objects = [""] + self.valid_key_value_pairs[self.key]
+
+    def is_empty(self):
+        return self.key == "" and self.operator == "" and self.value == ""
+
+    def validate(self):
+        if self.key == "":
+            self.set_css(self.key_select, valid=False)
+        else:
+            self.set_css(self.key_select)
+
+        if self.operator == "":
+            self.set_css(self.operator_select, valid=False)
+        else:
+            self.set_css(self.operator_select)
+
+        if self.value == "":
+            self.set_css(self.value_select, valid=False)
+        else:
+            self.set_css(self.value_select)
+
+        # if any attribute is "" then return False
+        return not (self.key == "" or self.operator == "" or self.value == "")
+
+    def construct_metadata_filter(self):
+        # probably not required, but just in case
+        if self.key == "" or self.operator == "" or self.value == "":
+            return None
+
+        return MetadataFilter(MetadataOperator[self.operator], self.key, self.value)
+
+    def __panel__(self):
+        return pn.Row(
+            self.key_select,
+            self.operator_select,
+            self.value_select,
+            self.delete_button,
+        )
+
+
+class MetadataFiltersBuilder(pn.viewable.Viewer):
+    filter_rows = param.List([])
+    corpus_names = param.List([])
+
+    def __init__(self, corpus_metadata, **params):
+        super().__init__(**params)
+
+        self.corpus_names_select = pn.widgets.Select(
+            options=[""] + self.corpus_names,
+            name="",
+            css_classes=["metadata-filter-row", "metadata-filter-key"],
+        )
+
+        self.corpus_metadata = corpus_metadata
+
+        self.add_filter_row_button = pn.widgets.ButtonIcon(
+            icon="circle-plus", width=25, height=25
+        )
+        self.add_filter_row_button.on_click(self.did_click_add_filter_row_button)
+
+        self.delete_buttons = []
+
+        self.filter_rows = [
+            FilterRow(
+                key="",
+                operator="",
+                value="",
+                on_delete_callback=self.delete_filter_row,
+            )
+        ]
+
+        self.corpus_names_select.param.watch(self.activate_filter_rows, "value")
+
+    def activate_filter_rows(self, event):
+        if event.new != "":
+            self.filter_rows = [
+                FilterRow(
+                    valid_key_value_pairs=self.corpus_metadata[
+                        self.corpus_names_select.value
+                    ],
+                    key="",
+                    operator="",
+                    value="",
+                    on_delete_callback=self.delete_filter_row,
+                )
+            ]
+
+    def did_click_add_filter_row_button(self, event):
+        if not self.filter_rows[-1].is_empty():
```

Comment:
Only add a new row if there is at least one non-empty element in the previous row.